File: py/bank.py
from money import Money


class Bank:
    def __init__(self):
        self.exchangeRates = {}

    def addExchangeRate(self, fromCurrency, toCurrency, rate):
        conversionKey = "%s->%s" % (fromCurrency, toCurrency)
        self.exchangeRates[conversionKey] = rate

    def convert(self, money, currency):
        if money.currency == currency:
            return Money(money.amount, currency)

        conversionKey = "%s->%s" % (money.currency, currency)
        if conversionKey not in self.exchangeRates:
            raise Exception(conversionKey)

        convertedAmount = money.amount * self.exchangeRates[conversionKey]
        return Money(convertedAmount, currency)

File: js/bank.js
const Money = require("./money");

class Bank {
    
    constructor() {
        this.exchangeRates = new Map();
    }

    addExchangeRate(from, to, rate) {
        const conversionKey = this.generateConversionKey(from, to);
        this.exchangeRates.set(conversionKey, rate);
    }

    convert(money, currency) {
        if (money.currency === currency) {
            return new Money(money.amount, money.currency);
        }

        const conversionKey = this.generateConversionKey(money.currency, currency);
        const rate = this.exchangeRates.get(conversionKey);
        if (rate === undefined) {
            throw Error(conversionKey);
        }
        return new Money(money.amount * rate, currency);
    }

    generateConversionKey = (from,to) => `${from}->${to}`;
}

module.exports = Bank;